src/controllers/ctlra/ctlracontroller.cpp

```diff
@@ -0,0 +1,147 @@
+/**
+  * @file ctlracontroller.cpp
+  * @author Harry van Haaren harryhaaren@gmail.com
+  * @date Thu Dec 22 2016
+  * @brief Ctlra backend implementation
+  *
+  */
+
+#include <wchar.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#include "controllers/ctlra/ctlracontroller.h"
+#include "util/path.h" // for PATH_MAX on Windows
+#include "controllers/defs_controllers.h"
+#include "util/compatibility.h"
+#include "util/trace.h"
+#include "controllers/controllerdebug.h"
+#include "util/time.h"
+
+#include "control/controlobject.h"
+#include "control/controlproxy.h"
+
+void CtlraController::event_func(struct ctlra_dev_t* dev,
+				 uint32_t num_events,
+				 struct ctlra_event_t** events)
+{
+	static const char* grid_pressed[] = { " X ", "   " };
+	struct ctlra_dev_info_t info;
+	ctlra_dev_get_info(dev, &info);
+
+	for(uint32_t i = 0; i < num_events; i++) {
+		struct ctlra_event_t *e = events[i];
+		const char *pressed = 0;
+		const char *name = 0;
+		switch(e->type) {
+		case CTLRA_EVENT_BUTTON:
+			name = ctlra_info_get_name(&info, CTLRA_EVENT_BUTTON,
+			                           e->button.id);
+			printf("[%s] button %s (%d)\n",
+			       e->button.pressed ? " X " : "   ",
+			       name, e->button.id);
+			break;
+
+		case CTLRA_EVENT_ENCODER:
+			name = ctlra_info_get_name(&info, CTLRA_EVENT_ENCODER,
+			                           e->encoder.id);
+			printf("[%s] encoder %s (%d)\n",
+			       e->encoder.delta > 0 ? " ->" : "<- ",
+			       name, e->button.id);
+			break;
+
+		case CTLRA_EVENT_SLIDER:
+			name = ctlra_info_get_name(&info, CTLRA_EVENT_SLIDER,
+			                           e->slider.id);
+			printf("[%03d] slider %s (%d)\n",
+			       (int)(e->slider.value * 100.f),
+			       name, e->slider.id);
+			if(e->slider.id == 0) {
+				// TODO: optimize to avoid repeated
+				// lookups of ControlProxy
+				ControlProxy("[Master]", "crossfader")
```

Comment:
I think it would make sense to keep the hash of Ctlra events to input callbacks on the C++ side to have a more consistent API with [how output callbacks are managed](https://mixxx.org/wiki/doku.php/midi_scripting#connect_output_callback_functions).